### ISSUE TITLE
Reduce response times for organization queues with many tasks

### ISSUE THREAD

**Issue by lowellrex** [department-of-veterans-affairs/caseflow#10170]
_This issue was forked from [a comment on a related ticket](https://github.com/department-of-veterans-affairs/caseflow/issues/9236#issuecomment-476719461)._

The VLJ support staff queue does not load cases because it takes [more than 5 minutes](https://github.com/department-of-veterans-affairs/caseflow/blob/cf6a53532a6b57a7c3bc58ee0d210fdb1b2a2905/client/app/queue/OrganizationQueueLoadingScreen.jsx#L20) for the server to respond to the browser's request for active tasks assigned to the organization (there are currently more than 2,000 active tasks assigned to the VLJ support staff organization). I was able to determine this by simulating the back-end component of this request as follows:

```ruby
rails c> u = Colocated.singleton.users.first
rails c> tasks = GenericQueue.new(user: Colocated.singleton).tasks
rails c> tasks.count
# 2086
rails c> js_tasks = ActiveModelSerializers::SerializableResource.new(
    AppealRepository.eager_load_legacy_appeals_for_tasks(tasks),
    user: u
  ).as_json
```

Below are a chart and table that show how long it takes to execute each phase of that last command which represents the data fetching part of the VLJ support staff organization page load (there are other components of that request, but this is almost certainly the most expensive). I ran this command twice this morning (10:20am eastern and 11:00am) from the production rails console so it is not a precise measurement of the time it takes to get a response to the browser, but it is indicative of where the bottlenecks are in that process.

![image](https://user-images.githubusercontent.com/32683958/55009692-61ced600-4fb9-11e9-9ebb-660b18797de6.png)

Phase | Trial 1 | Trial 2
-- | -- | --
Startup | 14 | 13
Initial VACOLS load (cases and tasks) | 58 | 62
VACOLS::Staff table load | 8 | 10
AttorneyCaseReviews load for serializer | 40 | 31
AvailableHearingLocations load for serializer | 60 | 48
Total time | 180 | 164

The most expensive phases of the data fetching part of the page load are 1) the initial VACOLS load of cases and tasks, and 2) the individual select statements on [the `AttorneyCaseReviews`](https://github.com/department-of-veterans-affairs/caseflow/blob/cf6a53532a6b57a7c3bc58ee0d210fdb1b2a2905/app/models/serializers/work_queue/task_serializer.rb#L99) and [`AvailableHearingLocations` tables](https://github.com/department-of-veterans-affairs/caseflow/blob/cf6a53532a6b57a7c3bc58ee0d210fdb1b2a2905/app/models/serializers/work_queue/task_serializer.rb#L79) while serializing fields for individual tasks. These two phases (or collection of phases) account for roughly half of the entire execution time.

Using this data we can speculate on ways to reduce the amount of time it takes for the server to respond, allowing the VLJ support staff organization queue to actually load. 
1. **Paginate the request**. The reason we are running into this problem for the VLJ support staff org queue but not other queues (except for hearings queue which we'll get to in a minute) is because of the sheer number of tasks in the queue. If we were to request only the first, say, 100 tasks we could dramatically reduce the amount of time both the serialization and eager VACOLS load phases take. Given the way we've designed other components of the system, this is probably extraordinarily complex though I imagine it will be needed at some point down the line if not now.
1. **Stop loading `AttorneyCaseReviews` and `AvailableHearingLocations` information in the serializer**. I don't know if this is feasible from a product perspective (haven't yet dug into where this information is used on the front-end) but this is almost certainly the quickest fix.
1. **Return summary task info on queue table views and detailed task info on case details pages**. We do something similar for appeals/appealDetails so this is a pattern we already have. This is certainly possible and maybe not all that complex, but it would require a fairly sizable refactor of the front-end (though maybe our increased reliance on reducers makes this refactor easier?).

There are some hybrid approaches here (for example, [defining `available_hearing_locations`](https://github.com/department-of-veterans-affairs/caseflow/blob/cf6a53532a6b57a7c3bc58ee0d210fdb1b2a2905/app/models/serializers/work_queue/task_serializer.rb#L79) as an "extra field" and [excluding extra fields from being serialized](https://github.com/department-of-veterans-affairs/caseflow/blob/cf6a53532a6b57a7c3bc58ee0d210fdb1b2a2905/app/controllers/organizations/tasks_controller.rb#L10) when there are many tasks: `exclude_extra_fields: tasks.count > 100`), but these are the distinct approaches I could come up with. Curious to hear if folks have thoughts around different approaches, hybrids, or preferences for which approach to pursue.

**Comment by monfresh** [comment]
Option 3 sounds reasonable to me. I've used it before and it is common for APIs to return different amounts of data depending on the context. I've also seen APIs that allow the requestor to specify which fields they want to be returned.

One thing on my to-do list as part of the performance improvement work I'm doing is to replace Active Model Serializers with Netflix's [fast_jsonapi](https://github.com/Netflix/fast_jsonapi), which is at least 25x faster.

**Comment by monfresh** [comment]
The fast_jsonapi gem makes it easy to specify only certain fields, without having to create a new serializer: https://github.com/Netflix/fast_jsonapi#sparse-fieldsets. I'm not sure if AMS supports that.

**Comment by monfresh** [comment]
I've been trying `fast_jsonapi` today and so far, the migration is going swimmingly. I should have a PR ready later tonight.

**Comment by laurjpeterson** [comment]
@lowellrex @hschallhorn @stwards11 - i see this one is still in current sprint column, but wonder if this ticket should actually be closed given the work already done to improve response times. should this ticket be closed, and if there is any additional work to be done, we create other tickets for it?

**Comment by hschallhorn** [comment]
Closed as #13714 will handle this for all orgs
